homeassistant/components/somfy/config_flow.py

```diff
@@ -0,0 +1,187 @@
+            description_placeholders={'authorization_url': url},
+            errors=errors,
+        )
+
+    async def _get_authorization_url(self):
+        """Get Somfy authorization url."""
+        from pymfy.api.somfy_api import SomfyApi
+        flow = self.hass.data[DATA_FLOW_IMPL][self.flow_impl]
+        client_id = flow[CLIENT_ID]
+        client_secret = flow[CLIENT_SECRET]
+        redirect_uri = '{}{}'.format(
+            self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
+        api = SomfyApi(client_id, client_secret, redirect_uri)
+
+        self.hass.http.register_view(SomfyAuthCallbackView())
+        return api.get_authorization_url()
+
+    async def async_step_code(self, code=None):
+        """Received code for authentication."""
+        if self.hass.config_entries.async_entries(DOMAIN):
+            return self.async_abort(reason='already_setup')
+
+        if code is None:
+            return self.async_abort(reason='no_code')
+
+        _LOGGER.debug("Should close all flows below %s",
+                      self.hass.config_entries.flow.async_progress())
+        # Remove notification if no other discovery config entries in progress
+
+        return await self._async_create_session(code)
+
+    async def _async_create_session(self, code):
+        """Create Somfy api and entries."""
+        flow = self.hass.data[DATA_FLOW_IMPL][DOMAIN]
+        client_id = flow[CLIENT_ID]
+        client_secret = flow[CLIENT_SECRET]
+        from pymfy.api.somfy_api import SomfyApi
+        redirect_uri = '{}{}'.format(
+            self.hass.config.api.base_url, AUTH_CALLBACK_PATH)
+        api = SomfyApi(client_id, client_secret, redirect_uri)
+        token = await self.hass.async_add_executor_job(
+            api.request_token, None, code)
+        _LOGGER.debug("Got new token")
+        # if not api.is_authorized:
+        #    _LOGGER.error('Authentication Error')
+        #    return self.async_abort(reason='auth_error')
+
+        _LOGGER.info('Successfully authenticated Somfy')
+        return self.async_create_entry(
+            title='',
+            data={
+                'token': token,
+                'refresh_args': {
+                    'client_id': client_id,
+                    'client_secret': client_secret
+                }
+            },
+        )
+
+
+class SomfyAuthCallbackView(HomeAssistantView):
+    """Somfy Authorization Callback View."""
+
+    requires_auth = False
+    url = AUTH_CALLBACK_PATH
+    name = AUTH_CALLBACK_NAME
+
+    @staticmethod
+    async def get(request):
+        """Receive authorization code."""
+        from aiohttp import web
+        hass = request.app['hass']
+        if 'code' in request.query:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={'source': 'code'},
+                    data=request.query['code'],
+                ))
+        return web.HTTPFound('/')
```

Comment:
@MartinHjelmare I don't know if it's the best soluton to redirect the end user to the HASS home page after Oauth process has been done. What do you think?

Reply:
I don't know about this.

Reply:
@MartinHjelmare Any idea on who can help on this topic?

Reply:
@balloob

Reply:
```python
        return web.Response(status=302, headers={
            'location': location
        })
```

Reply:
@balloob Sorry, if I was not clear. I don't want to redirect the end user. I've explained more in detail in this [comment](https://github.com/home-assistant/home-assistant/pull/19548#discussion_r260183137).